src/shogun/mathematics/linalg/ratapprox/logdet/LogDetEstimator.cpp

```diff
@@ -148,72 +125,81 @@ SGVector<float64_t> CLogDetEstimator::sample(index_t num_estimates)
-		for (index_t j=0; j<num_trace_samples; ++j)
-		{
-			SG_INFO("Computing log-determinant trace sample %d/%d\n", j,
-					num_trace_samples);
 
-			SG_DEBUG("Creating job for estimate %d, trace sample %d/%d\n", i, j,
+#ifdef HAVE_OPENMP // HAVE OPENMP
+#pragma omp single
+		{
+			num_threads = omp_get_num_threads();
+			step = num_vectors / num_threads;
+			num_threads--;
+		}
+		int32_t thread_num = omp_get_thread_num();
+#else
+		num_threads = 0;
+		step = num_vectors;
+		int32_t thread_num = 0;
+#endif
+		int32_t start = thread_num * step;
+		int32_t end =
+			(thread_num == num_threads) ? num_vectors : (thread_num + 1) * step;
+
+#ifdef WIN32 // HAVE WIN32
+		for (int32_t vec = start; vec < end; vec++)
+#else
+		for (int32_t vec = start; vec < end; vec++)
+#endif
+		{
+			pb.print_progress();
+			for (index_t j = 0; j < num_trace_samples; ++j)
+			{
+				SG_INFO(
+					"Computing log-determinant trace sample %d/%d\n", j,
 					num_trace_samples);
-			// get the trace sampler vector
-			SGVector<float64_t> s=m_trace_sampler->sample(j);
-			// create jobs with the sample vector and store the aggregator
-			CJobResultAggregator* agg=m_operator_log->submit_jobs(s);
-			aggregators->append_element(agg);
-			SG_UNREF(agg);
+				// get the trace sampler vector
+				SGVector<float64_t> s = m_trace_sampler->sample(j);
+				// calculate the result for sample s
+				float64_t agg = m_operator_log->solve(s);
+#pragma omp critical // so that the dynamic array stays concurrent
+				{
+					aggregators.append_element(agg);
+				}
+			}
 		}
 	}
-
-	REQUIRE(m_computation_engine, "Computation engine is NULL\n");
-
-	// wait for all the jobs to be completed
-	SG_INFO("Waiting for jobs to finish\n");
-	m_computation_engine->wait_for_all();
-	SG_INFO("All jobs finished, aggregating results\n");
+	// wait for all the computations to be completed
+	SG_INFO("Waiting for computations to finish\n");
+#pragma omp barrier
+	SG_INFO("All computations finished\n");
 
 	// the samples vector which stores the estimates with averaging
 	SGVector<float64_t> samples(num_estimates);
 	samples.zero();
 
-	// use the aggregators to find the final result
-	// use the same order as job submission to combine results
-	int32_t num_aggregates=aggregators->get_num_elements();
+	// preparing the final result
+	int32_t num_aggregates = aggregators.get_num_elements();
 	index_t idx_row=0;
 	index_t idx_col=0;
 	for (int32_t i=0; i<num_aggregates; ++i)
 	{
-		// this cast is safe due to above way of building the array
-		CJobResultAggregator* agg=dynamic_cast<CJobResultAggregator*>
-			(aggregators->get_element(i));
-		ASSERT(agg);
-
-		// call finalize on all the aggregators, cast is safe again
-		agg->finalize();
-		CScalarResult<float64_t>* r=dynamic_cast<CScalarResult<float64_t>*>
-			(agg->get_final_result());
-		ASSERT(r);
-
-		// iterate through indices, group results in the same way as jobs
-		samples[idx_col]+=r->get_result();
+		samples[idx_col] += aggregators.get_element(i);
```

Comment:
why don't you remove the `aggregators` variable and directly store in `samples`?

Reply:
i will use samples directly in the next patch